File: scripts/join_events_data.py
import json
import os


def join_events(output_file):
    """
    Joins multiple event data files into a single JSON file.

    Args:
        output_file (str): The path to the output JSON file.
    """

    events_data = []
    statuses = {"drop", "reroll", "completed"}

    for filename in os.listdir("events_data"):
        if filename.endswith(".json"):
            with open(
                os.path.join("events_data", filename), "r", encoding="utf-8"
            ) as file:
                data = json.load(file)
                data_games = data.get("games", [])
                filtered_games = [
                    game for game in data_games if game["completion_status"] in statuses
                ]
                events_data.extend(filtered_games)

    with open(output_file, "w", encoding="utf-8") as file:
        json.dump({"games": events_data}, file, ensure_ascii=False, indent=2)


if __name__ == "__main__":
    join_events("games_history.json")

File: src/join_events_data.ts
import { readdir } from "node:fs/promises";
import { join } from "node:path";
import { type CompletionStatus, type HistoryGame, GamesHistory } from "./types";
import { type as arktype } from "arktype";

const SCHEMA_FILE = "games_history_schema.json";

async function joinEventsData(outputFile: string): Promise<void> {
	const eventsData: HistoryGame[] = [];
	const statuses = new Set<CompletionStatus>(["drop", "reroll", "completed"]);

	const files = await readdir("events_data");
	files.sort();

	for (const filename of files) {
		if (!filename.endsWith(".json")) {
			continue;
		}

		console.log(`Processing ${filename}`);

		const data = await Bun.file(join("events_data", filename)).json();
		const parsedData = GamesHistory(data);

		if (parsedData instanceof arktype.errors) {
			console.error(
				`Error parsing ${filename}: ${parsedData.summary}, ${parsedData.issues}`,
			);
			continue;
		}

		const filteredGames = (parsedData.games ?? []).filter((game) =>
			statuses.has(game.completion_status),
		);

		eventsData.push(...filteredGames);
	}

	await Bun.write(
		outputFile,
		JSON.stringify(
			{
				"$schema": `https://raw.githubusercontent.com/eventlab-dev/games-history/refs/heads/main/${SCHEMA_FILE}`,
				games: eventsData,
			},
			null,
			2,
		),
	);
}

async function genSchemaFile() {
	const schema = GamesHistory.toJsonSchema();
	await Bun.write(SCHEMA_FILE, JSON.stringify(schema, null, 2));
}

await joinEventsData("games_history.json");
await genSchemaFile();